### ISSUE TITLE
Rendering SVGs with custom fonts on macOS

### ISSUE THREAD

**Issue by gfmio** [lovell/sharp#2399]
**What are you trying to achieve?**

Render SVGs with custom fonts on macOS

**Have you searched for similar questions?**

Yes, and I've found the ones covering the usage of fonts.conf on Linux / in AWS Lambda.

I followed the same method and when running the code in a linux container, fontconfig gets updated correctly, logs the debug output and renders the custom fonts.

On macOS, the custom fonts do not get picked up, fontconfig does not give any debug output and the default fonts get rendered.

Curioiusly, on both macOS and Linux, the command shelljs command does list the custom fonts, so it seems to me that fontconfig does pick up the correct settings, but then the rendering step doesn't make use of the updated config.

Does sharp / librsvg not use fontconfig on macOS? Or is there something else I need to configure? If so, what?

**Are you able to provide a minimal, standalone code sample that demonstrates this question?**

```ts
import * as fs from "fs";
import * as path from "path";
import sharp from "sharp";
import * as shell from "shelljs";

const fontConfigTemplate = (fontPath: string) => `<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<fontconfig>
  <dir>${fontPath}</dir>
  <cachedir>/tmp/fonts-cache/</cachedir>
  <config></config>
</fontconfig>
`;

const enableCustomFonts = () => {
  // Create the font config file
  const fontDir = path.resolve(path.join(__dirname, "assets", "fonts"));
  // console.log(fontDir);
  fs.writeFileSync(
    path.join(fontDir, "fonts.conf"),
    fontConfigTemplate(fontDir)
  );
  // Set the environment variable path
  process.env.FONTCONFIG_PATH = fontDir;
  // Set font config debugging to true
  process.env.FC_DEBUG = "1";
  console.log(shell.exec("fc-list").stdout);
};

export const main = async () => {
  enableCustomFonts();
  const svgPath = path.resolve(
    path.join(__dirname, "assets", "svgs", "MyCatIsGrumpy 2.svg")
  );

  // Render to image
  let image = sharp(svgPath, {
    density: 720
  });
  image = image.png();

  const outputBuffer = await image.toBuffer();

  fs.writeFileSync("out.png", outputBuffer);
};

main();
```

**Are you able to provide a sample image that helps explain the question?**

The test image I'm rendering is:

```svg
<svg viewBox="0 0 240 80" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
  <style type="text/css">
    .small { font: 13px; }
    .heavy { font: bold 30px; }
    .Rrrrr { font: 40px 'Mada'; fill: red; }
  </style>

  <rect x="120" y="10" width="50" height="50" rx="5" />
  <text x="20" y="35" class="small" style="font-family: 'Mada';">My</text>
  <text x="40" y="35" class="heavy" style="font-family: 'Decimal';">cat</text>
  <text x="55" y="55" class="small" style="font-family: 'Mercury Text G1';">is</text>
  <text x="65" y="55" class="Rrrrr" style="font-family: 'Brother-1816';">Grumpy!</text>
</svg>
```

The output on Linux:

![out](https://user-images.githubusercontent.com/2878342/95078448-126c7880-070d-11eb-90ae-516d983038d9.png)

The output on macOS:

![out-macos](https://user-images.githubusercontent.com/2878342/95078583-45af0780-070d-11eb-8fff-6c64ef74e5ce.png)

**Comment by lovell** [comment]
Have you tried setting the fontconfig-related environment variables before the call to `require('sharp')`?

The `FC_DEBUG` environment variable, e.g. `FC_DEBUG=1024`, might also help:
https://www.freedesktop.org/software/fontconfig/fontconfig-user.html#DEBUG

**Comment by gfmio** [comment]
Yes, I've tried using FC_DEBUG=1024 and setting the environment variables manually before even starting the node process.

When I manually invoke `FC_DEBUG=1024 FONTCONFIG_PATH=<PATH_TO_MY_CONFIG> FONTCONFIG_FILE=<PATH_TO_MY_CONFIG> fc-list`, it uses the correct config files and lists only my custom fonts.

When invoking `fc-list` using shelljs from my node script after setting the environment variables, it also uses the correct config files and lists only my custom fonts.

But when rendering the SVG with sharp, it nevertheless doesn't use the custom fonts.

**Comment by lovell** [comment]
The prebuilt binaries for macOS configure fontconfig with a default location of `/usr/local/etc` (so it will look for e.g. `/usr/local/etc/fonts/fonts.conf`), which is the same location as homebrew.

https://github.com/lovell/sharp-libvips/blob/master/build/lin.sh#L310

If you already have fontconfig installed via homebrew then you might need to run `fc-cache` to update its (shared) cache.

**Comment by gfmio** [comment]
I still haven't been able to get this to work, but I'll try with a new minimal example from scratch this weekend.

**Comment by lovell** [comment]
Thanks for the update, once you work it all out this is probably someone to add to the docs.

**Comment by cleversprocket** [comment]
I'm having the same problem. I'm on macOS Catalina 10.15.7. I checked and I have fonts.conf in the path specified above. Running `fc-list` in terminal I see all my installed fonts. I ran `fc-cache` but it didn't help. Admittedly I don't know anything about fontconfig.

**Comment by cleversprocket** [comment]
I got it to work. I was using `font-family: 'Suprema SemiBold';` which I verified is a valid font by viewing the SVG in the browser. I removed the second part of the font name `SemiBold` and sharp rendered the default Suprema font, which is `Suprema Regular` (or Suprema at 400 weight). I added `font-weight: 600;` to get the SemiBold font to render. 

Here's a trimmed down example:
```XML
<svg xmlns="http://www.w3.org/2000/svg">
  <style>
    .myClass {
      font-family: 'Suprema';
      font-weight: 600;
    }
  </style>
  <text class="myClass">Font renders correctly</text>
</svg>
```
I wish I knew why this works. @gfmio hopefully this will fix it for you, too.

**Comment by kleisauke** [comment]
If available, the `pango-list` and `pango-view` commands are also useful for debugging this. Sometimes the installed font could be named differently or is it necessary to separate the second part with a different delimiter. See for example:
```
$ pango-list | grep DIN
D-DIN Condensed 
  Regular:           D-DIN Condensed, Condensed
  DINCondensed-Bold: D-DIN Condensed, Bold Condensed
  *Italic:           D-DIN Condensed, Italic
  *Bold Italic:      D-DIN Condensed, Bold Italic
$ pango-view --font 'D-DIN Condensed Bold' --no-display --text 'ABCDEFGHIJKLMNOPQRSTUVWXYZ' --output out.png 
$ pango-view --font 'D-DIN Condensed, Bold' --no-display --text 'ABCDEFGHIJKLMNOPQRSTUVWXYZ' --output out2.png 
```

Output:
![out](https://user-images.githubusercontent.com/12746591/96850054-aac96380-1456-11eb-8115-5e2bbdd65db5.png)
![out2](https://user-images.githubusercontent.com/12746591/96850062-ad2bbd80-1456-11eb-8fa6-4b67470fcb75.png)

**Comment by kleisauke** [comment]
> Does sharp / librsvg not use fontconfig on macOS?

Pango (the text renderer used by librsvg) defaults to the CoreText rendering and mapping backend on macOS. You could try to use the Fontconfig backend with the `PANGOCAIRO_BACKEND="fontconfig"` environment variable.

**Comment by lovell** [comment]
Great sleuthing Kleis, I did not know this. Perhaps sharp should automagically set this env variable when using the prebuilt binaries?

**Comment by kleisauke** [comment]
That could work, Alternatively, we might be able to build Pango without CoreText support by removing this:
https://github.com/GNOME/pango/blob/c7e6f369079f60311cf241e8e05f4cd92f7c344b/meson.build#L371-L384

(Note that I'm not sure what the advantages of CoreText are compared to Fontconfig and vice versa)

**Comment by lovell** [comment]
The env variable approach would allow the proposed new default setting of `PANGOCAIRO_BACKEND="fontconfig"` to be overridden back to CoreText via `PANGOCAIRO_BACKEND="coretext"` if required, so I'd be hesitant about removing this ability entirely.

**Comment by lovell** [comment]
Commit ef964b5 changes the default behaviour of `pango` to use `fontconfig`, which will be in v0.27.0.

If required, setting `PANGOCAIRO_BACKEND="coretext"` will restore the previous behaviour.

**Comment by lovell** [comment]
sharp v0.27.0 now available, thanks for reporting this.

**Comment by lovell** [comment]
sharp v0.27.1 will revert this change as it is causing too many problems - see #2515

If you require consistent cross-platform font behaviour, please continue to set `PANGOCAIRO_BACKEND="fontconfig"`

**Comment by GeoffreyBooth** [comment]
I’m trying to get Sharp to render SVGs on macOS without the fonts needing to be installed on the system, like the original post. I have it working within Docker per https://constantsolutions.dk/2022/11/how-to-install-custom-fonts-on-a-nodejs-image/, but adapting the code above and using `PANGOCAIRO_BACKEND="fontconfig"` crashes with

```
Pango-CRITICAL **: 23:02:15.997: Unknown $PANGOCAIRO_BACKEND value.
Available backends are: coretext
```

@lovell is there a new recommendation nowadays?

Here’s my full code:

```js
import { writeFile } from 'node:fs/promises'
import { platform, tmpdir } from 'node:os'
import { join } from 'node:path'
import { cwd, env } from 'node:process'
/** @import Sharp from 'sharp' */


/** @type {Sharp} */
let sharp
/**
 * For macOS, configure Sharp to use this app’s fonts without needing them installed in the system.
 * When running via the Docker container, the fonts are already installed in the image.
 * Windows and other platforms are not supported.
 * Adapted from https://github.com/lovell/sharp/issues/2399
 */
export async function getSharp() {
	if (sharp) {
		return sharp
	}

	if (platform() === 'darwin') {
		const fontPath = join(cwd(), 'assets', 'fonts')

		const fontConfig = `<?xml version="1.0"?>
	<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
	<fontconfig>
		<dir>${fontPath}</dir>
		<cachedir>${tmpdir()}/fonts-cache/</cachedir>
		<config></config>
	</fontconfig>
	`

		await writeFile(join(fontPath, 'fonts.conf'), fontConfig)
		env.FONTCONFIG_PATH = fontPath

		// Throws `Unknown $PANGOCAIRO_BACKEND value. Available backends are: coretext`
		// env.PANGOCAIRO_BACKEND = 'fontconfig'

		// For debugging
		// await import('node:child_process').then(({ execSync }) => {
		// 	console.log(execSync('fc-list').toString())
		// })
	}

	// All of the above needs to happen before Sharp is imported
	sharp = (await import('sharp')).default

	return sharp
}
```

**Comment by lovell** [comment]
@GeoffreyBooth Are you using prebuilt binaries provided by sharp or a globally-installed `vips` e.g. via Homebrew? Are you using the latest version of everything involved? Is this on an Intel or ARM CPU? This might be a new issue as there were some recent changes to the way FreeType and Cairo are being built.

**Comment by lovell** [comment]
@GeoffreyBooth I've updated the script that creates the prebuilt libvips binaries provided by sharp via commits https://github.com/lovell/sharp-libvips/commit/27c642811bb69264d43eedee1833c1ac8927fd6f and https://github.com/lovell/sharp-libvips/commit/273d659a1bd84ba554678bf93d76140de800d34c to ensure `PANGOCAIRO_BACKEND=fontconfig` will work on macOS. You'll have to wait until the next release of libvips to take advantage of this, probably v8.16.0.

For now, as a possible workaround, installing `vips` via Homebrew should support this env var according to https://github.com/Homebrew/homebrew-core/blob/6c8204a896bf1243537949f5de67f3201cdf636f/Formula/c/cairo.rb#L47-L48
